File: java-gdb-activities-20/activity20/src/com/gdb/logging/DatabaseLogDestination.java
package com.gdb.logging;

import com.gdb.command.TransactionCommand;
import com.gdb.db.SimulatedDatabase;
import java.util.*;
import java.util.stream.Collectors;

public class DatabaseLogDestination implements LogDestination {
    private static final String TABLE = "transaction_log";

    private final SimulatedDatabase db;

    public DatabaseLogDestination(SimulatedDatabase db) {
        this.db = db;
    }

    @Override
    public void write(TransactionCommand cmd) {
        if (db != null) {
            db.insert(TABLE, cmd);
        }
    }

    @Override
    public List<TransactionCommand> readAll() {
        if (db == null) return Collections.emptyList();
        return db.selectAll(TABLE).stream()
                .map(o -> (TransactionCommand) o)
                .collect(Collectors.toList());
    }

    @Override
    public void clear() {
        if (db != null) {
            db.deleteAll(TABLE);
        }
    }

    @Override
    public String getDestinationName() {
        return "DATABASE";
    }
}

File: python-gdb-activities-20/activity18-solution/gdb/logging/database_log_destination.py
# gdb/logging/database_log_destination.py
from typing import List
from gdb.domain.transaction import Transaction
from gdb.logging.log_destination import LogDestination
from gdb.db.simulated_database import SimulatedDatabase

class DatabaseLogDestination(LogDestination):
    """Database storage destination implementor backed by SimulatedDatabase."""

    def __init__(self, db: SimulatedDatabase) -> None:
        self._db = db

    def write(self, transaction: Transaction) -> None:
        self._db.insert(transaction)

    def read_all(self) -> List[Transaction]:
        return self._db.query_all()

    def get_destination_name(self) -> str:
        return "Database"
